### PR TITLE
Notebook

### PR DESCRIPTION
**Pull request by DerekLiu35**
created notebook of code

### CODE REVIEW

**Comment by thomas-reimonn** [comment]
Feedback from meeting:
- Fix selection of negatives (overlap, left join, remove bins without an overlap)
- Plot loss across epochs
- Plot convolutional kernel weights using logomaker
- Pick "best" kernel and sweep it across positive sequences
- Plot the kernel * sequence score across the strand (pick a random positive for this)
- Select the sub_sequence that matches the kernel with a MAX value (for all positives)
- Average & plot the sub_sequences

**Comment by thomas-reimonn** [comment]
Example of using logomaker: 
```
    def plot_kernel_weights(self, pdf):
        # Plot kernel weights
        fig, ax = plt.subplots(3, 3)
        model: tf.keras.Model = self._model
        convolution = model.get_layer("convolution")
        weights = convolution.get_weights()
        for i in range(3):
            for j in range(3):
                w = weights[0][:, :, i * 3 + j]
                df = pd.DataFrame(w)
                df.columns = ["A", "C", "G", "T"]
                logger.debug(f"Creating logo and plotting kernel weights: \n{df}\n")
                logomaker.Logo(df, ax=ax[i, j])
                ax[i, j].set_title(f"Kernel {i*3+j} weights")
        fig.savefig(pdf, format="pdf")
```

**Comment by thomas-reimonn** [comment]
Sometimes the position probability matrix of information content matrix is easier to visualize than the position weight matrix. Here is an example of computing it:
```
def information_content(pwm):
    """
    Compute the information content of a Position Weight Matrix (PWM).

    Parameters:
        pwm (numpy.ndarray): A 2D numpy array representing the PWM.
                             Each row corresponds to a nucleotide (A, C, G, T),
                             and each column corresponds to a position in the motif.

    Returns:
        numpy.ndarray: An array containing the information content of each position in the PWM.
    """
    pseudo_count = 0.001
    num_positions = pwm.shape[1]
    information_content_arr = np.zeros(num_positions)

    for pos in range(num_positions):
        col_sum = np.sum(pwm[:, pos])
        for nucleotide in range(4):  # 4 for A, C, G, T
            prob = (pwm[nucleotide, pos] + pseudo_count) / (col_sum + 4 * pseudo_count)
            information_content_arr[pos] += prob * math.log2(prob / 0.25)

    return information_content_arr
```

**Comment by thomas-reimonn** [comment]
Another thing you can try is predicting on a different cell type. Search the ENCODE portal "encodeproject.org" for a CTCF TF ChIP-seq experiment in another cell type, find the IDR thresholded peaks bed file, download that and use that as your validation set.

**Comment by vedatonuryilmaz** [comment]
Hi @DerekLiu35 , can you also add Nezar to the repo? His gb username is @nvictus

**Comment by thomas-reimonn** [comment]
Tensorflow code for reverse complement model:
```
    def create_model(self) -> tf.keras.Model:
        """
        Create a tensorflow::keras classifier model
        """
        sequence_input = tf.keras.Input(shape=(self.width, 4))
        reverse_complement = tf.keras.layers.Lambda(lambda x: tf.keras.backend.reverse(x, axes=(1, 2)))
        reverse = tf.keras.layers.Lambda(lambda x: tf.keras.backend.reverse(x, axes=1))
        reverse_complement_sequence = reverse_complement(sequence_input)
        shared_conv = tf.keras.layers.Conv1D(
            filters=self.num_kernels,
            kernel_size=self.kernel_width,
            strides=1,
            padding="valid",
            activation="relu",
            use_bias=False,
            name="convolution",
            # kernel_regularizer=tf.keras.regularizers.l1(l1=0.0005),
            kernel_constraint=tf.keras.constraints.non_neg(),
        )
        forward_convultion = shared_conv(sequence_input)
        reverse_convultion = shared_conv(reverse_complement_sequence)
        reversed_features = reverse(reverse_convultion)
        merged = tf.keras.layers.Concatenate()([forward_convultion, reversed_features])
        pooled = tf.keras.layers.GlobalMaxPooling1D()(merged)
        dense1 = tf.keras.layers.Dense(16, activation="sigmoid")(pooled)
        dense2 = tf.keras.layers.Dense(16, activation="sigmoid")(dense1)

        output = tf.keras.layers.Dense(
            1,
            activation="sigmoid",
        )(dense2)
        model = tf.keras.Model(inputs=sequence_input, outputs=output)
        model.compile(
            loss="binary_crossentropy",
            metrics=[
                tf.keras.metrics.BinaryAccuracy(),
                tf.keras.metrics.AUC(
                    name="auroc",
                    num_thresholds=200,
                    curve="ROC",
                ),
                tf.keras.metrics.AUC(
                    name="aupr",
                    num_thresholds=200,
                    curve="PR",
                ),
            ],
            optimizer=tf.keras.optimizers.Adam(0.001),
        )
        logger.info("Created tensorflow model.")
        logger.info(f"Summary of model: {model.summary()}")
        return model```

**Comment by thomas-reimonn** [comment]
Summary of discussion from this morning:
- Reverse complement in biology
- Updating the model to compute convolutions for the forward and reverse complement strand, then concatenate the forward and reversed(convolutions) values
- Using 1D global max pool to increase generalizability and decrease compute time
- Using a validation set to track training history and avoid overfitting
- Creating a logo based on only the directional sequence -- possibly thresholding the max value and picking which strand to use....
- Should probably make two logos: forwards and reverse by thresholding the max value of the convolutional kernel

**Comment by thomas-reimonn** [comment]
From Nezar:
- Use bioframe set diff instead of overlap to find negatives
- Exclude ummappable bins from negative set, using a genome mappability annotation file
- Choice of negatives is very important in biology. A portion of the genome is unmappable or very variable. So choosing "relevant" negatives is important.
